### PR TITLE
Integrate cluster manager

### PR DESCRIPTION
**Pull request by lvrach**
## Description of the change

> Introduced a new way to start and manage the mode of the server. Now JobsDB is being created with the help of new JobsDB creators and managed with the help of jobsDB interface. Processor and Routers are also created by the respective creators and managed by the help of their respective lifecycle manager interfaces. 
JobsDBs along with Processor and Router are passed to Cluster Manager in order to manage (start/stop) individual components of the server based on the expected state of the server or to help the server reach to the desired state (Normal/Degraded Mode).
The cluster manager needs one mode provide also based on which it gets the signal of mode changes. 
server_LCM

Configs: 
LegacyAppHandler   EnableMT
True                          True       Older App handlers With New Pickup Algo (In-mem stats being populated)
False                         False      New AppHandler with Old Pickup Algo (In-mem stats being populated)
False                         True        New AppHandler with New Pickup algo (In-mem stats being populated) –Hosted/end-goal
True                          False       Older AppHandler with older pickup Algo (In-mem stats being populated) – Current Default

## Notion Link

> [Notion Link](https://www.notion.so/rudderstacks/Integration-of-Cluster-Manager-451b5da33c2944919996532b8d8d1a0e)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: app/apphandlers/processorAppHandler.go [hunk 1]
```diff
@@ -88,28 +89,33 @@ func (processor *ProcessorApp) StartRudderCore(ctx context.Context, options *app
...
+	defer brtDB.Close()
+	errDB = jobsdb.NewForRead("proc_error", jobsdb.WithRetention(routerDBRetention), jobsdb.WithMigrationMode(migrationMode),
+		jobsdb.WithStatusHandler(), jobsdb.WithQueryFilterKeys(jobsdb.QueryFiltersT{}))
+
+	var tenantRouterDB jobsdb.MultiTenantJobsDB = &jobsdb.MultiTenantLegacy{HandleT: rtDB} //FIXME copy locks ?
+	var multitenantStats multitenant.MultiTenantI = multitenant.NewStats(tenantRouterDB)
+
+	// TODO: remove if
+	// TODO: Think what we need to do for replay
 	if enableProcessor || enableReplay {
```
Comment (lvrach) [inline app/apphandlers/processorAppHandler.go:116]:
```suggestion
```

File: app/apphandlers/processorAppHandler.go [hunk 2]
```diff
@@ -121,48 +127,60 @@ func (processor *ProcessorApp) StartRudderCore(ctx context.Context, options *app
...
 					clearDB := false
-					StartProcessor(ctx, &clearDB, enableProcessor, &gatewayDB, &routerDB, &batchRouterDB, &procErrorDB, reportingI, multitenantStats)
+					StartProcessor(ctx, &clearDB, enableProcessor, gwDB, rtDB, brtDB, errDB, reportingI, multitenantStats)
 
 					return nil
 				})
 			}
 			startRouterFunc := func() {
 				g.Go(func() error {
-					StartRouter(ctx, enableRouter, tenantRouterDB, &batchRouterDB, &procErrorDB, reportingI, multitenantStats)
```
Comment (lvrach) [inline app/apphandlers/processorAppHandler.go:131]: @kuldeep0020 I find `batchRouterDB` more readable than `brtDB`, what do you think?

File: app/apphandlers/processorAppHandler.go [hunk 3]
```diff
@@ -88,81 +90,105 @@ func (processor *ProcessorApp) StartRudderCore(ctx context.Context, options *app
...
-	gatewayDB.Setup(jobsdb.Read, options.ClearDB, "gw", gwDBRetention, migrationMode, true, jobsdb.QueryFiltersT{})
-	defer gatewayDB.TearDown()
-
-	var routerDB jobsdb.HandleT = jobsdb.HandleT{}
+	reportingI := processor.App.Features().Reporting.GetReportingInstance()
 
-	var tenantRouterDB jobsdb.MultiTenantJobsDB = &jobsdb.MultiTenantLegacy{HandleT: &routerDB} //FIXME copy locks ?
+	//IMP NOTE: All the jobsdb setups must happen before migrator setup.
+	gwDB := jobsdb.NewForRead("gw", jobsdb.WithRetention(gwDBRetention), jobsdb.WithMigrationMode(migrationMode),
+		jobsdb.WithStatusHandler(), jobsdb.WithQueryFilterKeys(jobsdb.QueryFiltersT{}))
```
Comment (lvrach) [inline app/apphandlers/processorAppHandler.go:97]: [minor] for readability:

```suggestion
	gwDB := jobsdb.NewForRead(
	    "gw", 
	    jobsdb.WithRetention(gwDBRetention), 
	    jobsdb.WithMigrationMode(migrationMode),
	    jobsdb.WithStatusHandler(), 
	    jobsdb.WithQueryFilterKeys(jobsdb.QueryFiltersT{})
        )
```

Comment (kuldeep0020) [reply app/apphandlers/processorAppHandler.go]: done

File: app/apphandlers/processorAppHandler.go (hunk 1 above)
Comment (kuldeep0020) [reply app/apphandlers/processorAppHandler.go]: make sense, removing

File: app/apphandlers/processorAppHandler.go (hunk 2 above)
Comment (kuldeep0020) [reply app/apphandlers/processorAppHandler.go:131]: Okay 😓

**Comment by cisse21** [comment]
Is there any way to fall back to older implementation? If something goes wrong in production can we bring back the system to stability by tuning some config parameters?

**Comment by lvrach** [comment]
Good point we need to think about this a bit more .

We haven’t introduced any major logic, but we can think of an easy to
switch between the existing and the new flow.

File: app/apphandlers/embeddedAppHandler.go [hunk 4]
```diff
@@ -58,109 +62,172 @@ func (embedded *EmbeddedApp) StartRudderCore(ctx context.Context, options *app.O
...
 		var replayDB jobsdb.HandleT
 		replayDB.Setup(jobsdb.ReadWrite, options.ClearDB, "replay", routerDBRetention, migrationMode, true, jobsdb.QueryFiltersT{})
 		defer replayDB.TearDown()
-		embedded.App.Features().Replay.Setup(&replayDB, &gatewayDB, &routerDB, &batchRouterDB)
+		embedded.App.Features().Replay.Setup(&replayDB, gwDBForProcessor, routerDB, batchRouterDB)
 	}
 
 	if enableGateway {
-		var gateway gateway.HandleT
+		var gw gateway.HandleT
```
Comment (kuldeep0020) [inline app/apphandlers/embeddedAppHandler.go:204]: don't use separate var initialisation.

Comment (kuldeep0020) [reply app/apphandlers/embeddedAppHandler.go:204]: same thing for ratelimiter

File: app/apphandlers/gatewayAppHandler.go [hunk 5]
```diff
@@ -37,43 +36,49 @@ func (gatewayApp *GatewayApp) StartRudderCore(ctx context.Context, options *app.
...
-			gatewayApp.App.Features().Migrator.PrepareJobsdbsForImport(&gatewayDB, nil, nil)
+			gatewayApp.App.Features().Migrator.PrepareJobsdbsForImport(gatewayDB, nil, nil)
 		}
 	}
 
 	g, ctx := errgroup.WithContext(ctx)
 
 	if enableGateway {
-		var gateway gateway.HandleT
+		var gw gateway.HandleT
```
Comment (kuldeep0020) [inline app/apphandlers/gatewayAppHandler.go:69]: don't use separate initialiser for gw and ratelimiter

File: app/apphandlers/processorAppHandler.go [hunk 6]
```diff
@@ -88,81 +90,126 @@ func (processor *ProcessorApp) StartRudderCore(ctx context.Context, options *app
...
-		procErrorDB.Setup(jobsdb.ReadWrite, options.ClearDB, "proc_error", routerDBRetention, migrationMode, false, jobsdb.QueryFiltersT{})
-		defer procErrorDB.TearDown()
-
-		if config.GetBool("EnableMultitenancy", false) {
-			tenantRouterDB = &jobsdb.MultiTenantHandleT{HandleT: &routerDB}
-			multitenantStats = multitenant.NewStats(tenantRouterDB)
-		}
+	if config.GetBool("EnableMultitenancy", false) {
+		tenantRouterDB = &jobsdb.MultiTenantHandleT{HandleT: routerDB}
+		multitenantStats = multitenant.NewStats(tenantRouterDB)
```
Comment (kuldeep0020) [inline app/apphandlers/processorAppHandler.go:138]: Always enable multitanant stats

File: app/apphandlers/embeddedAppHandler.go [hunk 7]
```diff
@@ -58,109 +62,172 @@ func (embedded *EmbeddedApp) StartRudderCore(ctx context.Context, options *app.O
...
+		jobsdb.WithStatusHandler(),
+		jobsdb.WithQueryFilterKeys(jobsdb.QueryFiltersT{}),
+		)
+
+	// TODO: Always initialize multi-tenant stats after PR#1736 gets merged.
+	var tenantRouterDB jobsdb.MultiTenantJobsDB = &jobsdb.MultiTenantLegacy{HandleT: routerDB}
+	var multitenantStats multitenant.MultiTenantI = multitenant.NOOP
+	if config.GetBool("EnableMultitenancy", false) {
+		tenantRouterDB = &jobsdb.MultiTenantHandleT{HandleT: routerDB}
+		multitenantStats = multitenant.NewStats(tenantRouterDB)
```
Comment (kuldeep0020) [inline app/apphandlers/embeddedAppHandler.go:118]: Always enable multitanant stats

File: app/apphandlers/processorAppHandler.go [hunk 8]
```diff
@@ -88,81 +90,126 @@ func (processor *ProcessorApp) StartRudderCore(ctx context.Context, options *app
...
 	}
 
-	operationmanager.Setup(&gatewayDB, &routerDB, &batchRouterDB)
+	var modeProvider state.StaticProvider
+	// FIXME: hacky way to determine servermode
+	if enableProcessor && enableRouter {
+		modeProvider = state.StaticProvider{
+			Mode: servermode.NormalMode,
+		}
+	}
```
Comment (kuldeep0020) [inline app/apphandlers/processorAppHandler.go:174]: provide Degraded mode in else part

File: app/cluster/dynamic.go [hunk 9]
```diff
@@ -64,16 +63,25 @@ func (d *Dynamic) Setup()  {
...
 	for {
 		select {
 		case <-ctx.Done():
 			if d.currentMode == servermode.NormalMode {
 				d.stop()
 			}
 			return nil
-		case newMode := <-d.Provider.ServerMode():
+		case newMode := <-modeChan:
 			d.logger.Debugf("Got trigger to change the mode, new mode: %s, old mode: %s", newMode.Mode(), d.currentMode)
```
Comment (kuldeep0020) [inline app/cluster/dynamic.go:83]: Change it to info

File: app/cluster/dynamic.go [hunk 10]
```diff
@@ -64,16 +63,25 @@ func (d *Dynamic) Setup()  {
...
-			err := d.handleModeChange(newMode.Mode())
+			err = d.handleModeChange(newMode.Mode())
 			if err != nil {
 				d.logger.Error(err)
```
Comment (kuldeep0020) [inline app/cluster/dynamic.go:86]: remove this error logging

File: app/cluster/dynamic_test.go [hunk 11]
```diff
@@ -18,14 +19,17 @@ type mockModeProvider struct {
...
+func (m *mockModeProvider) ServerMode() (<-chan servermode.Ack, error) {
+	return m.ch, nil
 }
 
 func (m *mockModeProvider) SendMode(newMode servermode.Ack) {
 	m.ch <- newMode
 }
 
+func (m *mockModeProvider) Close() {
+
```
Comment (kuldeep0020) [inline app/cluster/dynamic_test.go:31]: close the channel

File: app/apphandlers/embeddedAppHandler.go (hunk 4 above)
Comment (kuldeep0020) [reply app/apphandlers/embeddedAppHandler.go:204]: done

File: app/apphandlers/gatewayAppHandler.go (hunk 5 above)
Comment (kuldeep0020) [reply app/apphandlers/gatewayAppHandler.go:69]: Done

File: app/apphandlers/processorAppHandler.go (hunk 6 above)
Comment (kuldeep0020) [reply app/apphandlers/processorAppHandler.go:138]: done

File: app/apphandlers/embeddedAppHandler.go (hunk 7 above)
Comment (kuldeep0020) [reply app/apphandlers/embeddedAppHandler.go:118]: done

File: app/apphandlers/processorAppHandler.go (hunk 8 above)
Comment (kuldeep0020) [reply app/apphandlers/processorAppHandler.go:174]: done

File: app/cluster/dynamic.go (hunk 9 above)
Comment (kuldeep0020) [reply app/cluster/dynamic.go:83]: done

File: app/cluster/dynamic.go (hunk 10 above)
Comment (kuldeep0020) [reply app/cluster/dynamic.go:86]: done

File: app/cluster/dynamic_test.go (hunk 11 above)
Comment (kuldeep0020) [reply app/cluster/dynamic_test.go:31]: done

**Review by cisse21 — APPROVED** [review]

**Pull request merged**